### DISCUSSION TITLE
More Code Typing Feature(Repo mode?)

### DISCUSSION THREAD

**Discussion by afairley** [monkeytypegame/monkeytype#5634] [Questions]
Hi,

I've been playing around with the free version of typing.io for a while and would like to implement a feature for monkeytype where you say put in a github url and then do typing tests based on code files from said git repo.  Theoretically this might include typing your way through all the code files in the repo if you are a very dilligent and committed and fast typist. 

Presumably this would be an additional mode similar to the Quotes mode, tenatively called the Repo mode, and would presumably require at least storing a few bits of per user data. I guess the bare bones way to do the data side would be a url for the git repository the user is working from, the specific commit at which they started at and a pointer into where in the text from the repository the user is. That's probably not a lot of additional storage per user.

I've just started reading the codebase and it looks very good. Looking at https://github.com/monkeytypegame/monkeytype/blob/master/CONTRIBUTING_ADVANCED.md , it looks like the style guide is mainly to run Prettier with the default config against my changes?

I can do a more serious feature proposal if you'd like, I'll be happy to do all the work of actually implementing the feature by myself, which isn't to say I don't mind help.

Cheers,

Alex

**Comment by kamykaze** [comment]
As someone who wants to get faster at typing to support my coding work, I wish this had more interest. But maybe this isn't very visible because it's marked as a 'question' instead of a 'feature' and most questions are users asking for help.
 
That being said, I think one thing worth discussing is symbols.

How would you approach the opening and closing of brackets, parenthesis, etc.? In a real world, I don't think too many coders will type '(' then type all the code that needs to go in there, then type ')' . You'd lose track of all your open symbols. I think most people either rely on their code editor to auto close their symbol, or they type both opening and closing symbols, then use the arrow keys to go back and insert more code.

Another big one...spaces and tabs.

In typing.io, they automatically put you in the right position so you don't have to tab when you go into a new line. However, in practice, your editor will only tab you to keep you aligned with the previous tabbing or the line above, not take you a nested level deeper.

These and other behaviors that only a code editor would normally handle for you would need to be added to this repo mode, or added as settings options so they can be used for a more realistic code typing experience.
